crowbar_framework/app/models/api/node.rb

```diff
@@ -16,6 +16,98 @@
 
 module Api
   class Node < Tableless
+    def initialize(name = nil)
+      @node = NodeObject.find_node_by_name name
+    end
+
+    def upgraded?
+      # FIXME: check this by looking at some upgraded-ok file on the node
+      current_platform = "#{@node[:platform]}-#{@node[:platform_version]}"
+      @upgraded ||= current_platform == @node[:target_platform]
+    end
+
+    def pre_upgrade
+      # TODO: save the global status info about this substep (we started prepare for upgrade)
+
+      # Migrate out the l3 agents and shut down pacemaker
+      script = "/usr/sbin/crowbar-pre-upgrade.sh"
+      out = @node.run_ssh_cmd(script)
```

Comment:
There *is* a good reason for `screen`, which is what I am trying to say :-)

Running the script again is a way of resuming it; they are not mutually exclusive things. I'm totally fine with this approach based on making the script idempotent, but that misses the point.  Even if each step is idempotent, it's not atomic.  For example, if the script gets killed due to a severed `ssh` connection whilst it's in the middle of database migrations, this might cause some consistency errors at the OpenStack level which could cause even an idempotent db_sync to get stuck on second time around.  I'm not sure if that's a good example, but the point is that if we can avoid non-atomic operations from being interrupted by a killed network connection then we should.